### ISSUE TITLE
pgsql RA times out on monitor operation

### ISSUE THREAD

**Issue by fjewl** [ClusterLabs/resource-agents#841]
Hi,
we now had two incidents in the last two days with Pacemaker triggering a demote on the master ressource because the pgsqsl RA ran into a time out situation (configured for 60secs) for the monitor op.

As far as I understand, the monitor op is implemented at the the end by pgsql_real_monitor. Looking at pgsql_real_monitor we don't see how this could run into an execution timeout apart from the SQL query (EDIT: defaulting to SELECT * FROM pg_is_in_recovery"). At all times during the fail-overs our master seemed to be doing just fine, there was no particular load issue as well. 

It would be great if some debugging log could be added (I might give it a try if you could point me to some documentation on how to correcly implement logging), so we can drill down on this time out so see where the RA is actually stuck.

Perhaps you also have some other idea why the pgsql_monitor process  could time out.

Thanks,
Ludwig

**Comment by kskmori** [comment]
Tracing RA feature would be helpful in such cases, which is not limited to pgsql RA.
See 'crm help resource trace' for details. (if you are using crmsh. I'm not sure how in pcs though)

Some random reasons that caused an unexpected timeout from my experiences (besides high loads) are: anti-virus software, which would defer or prevent to run system binaries during the scan, or 'su' command stuck, maybe something wrong in the PAM.

Hope it helps.

**Comment by fjewl** [comment]
Dear @kskmori, thanks for the feedback. I have been adding debug statements , but the trace operation seems indeed a better option, didn't know about that , thanks. 
We can most probably rule out anti virus. PAM had been an issue in the past, especially as we are using ldap authentication, however the user for executing the calls are configured to authenticate locally. Still, good remark, we are focusing on the sys calls rather than the database queries. I would like to keep this issue open until i can provide further insights and see if we can provide any potential improvements.

Fjrwl

**Comment by fjewl** [comment]
All. I am closing the issue as we cannot reproduce it unfortunately. We have added debug logging statements to the RA and analyzed the different calls to no avail. We cannot pinpoint the issue to any specific topic, potential candidates are still DNS and PAM issues during su. Also the database cluster in question is under quite some load (other clusters are not affected) so we suspect a combination / some hickups. We have now increased the timeout setting and plan to scale out the cluster itself.
